[Clang][AArch64] Generalise streaming mode checks for builtins.

sdesmalen-arm opened the pull request:
PR #76975 added 'IsStreamingOrSVE2p1' to emit a diagnostic when a builtin marked
with 'IsStreamingOrSVE2p1' is used in a non-streaming function that is not
compiled with `+sve2p1`.

The problem is a bit more complex than only this case. For example, we've marked
lots of builtins with 'IsStreamingCompatible', meaning it can be used in either
streaming, streaming-compatible or non-streaming functions. But the code in
SemaChecking, doesn't check the appropriate target guards. This issue becomes
relevant when SVE builtins are only available in streaming mode, e.g. when
compiling for SME without SVE.

If we were to add the appropriate target guards, we'd have to add many more
combinations, e.g.:
  IsStreamingSMEOrSVE
  IsStreamingSME2OrSVE2
  IsStreamingSMEOrSVE2p1
  IsStreamingSME2OrSVE2p1
  etc.

To avoid having to add more combinations (and avoid having to add more in the
future for new extensions), we use a single 'IsSVEOrStreamingSVE' flag for all
builtins that are available in streaming mode for the appropriate SME flags, or
in non-streaming mode for the appropriate SVE flags, or both.  The code in
SemaChecking will then verify for which mode (or both) the builtin would be
defined, given the target features of the function/compilation unit.

For example:

  'svclamp' is enabled under FEAT_SVE2p1 and FEAT_SME2

* When we compile for SVE2p1 and SME (but not SME2), the builtin is undefined
  behaviour when called from a streaming function.

* When we compile for SME2 and SVE2 (but not SVE2p1), the builtin is undefined
  behaviour when called from a non-streaming function.

* When we compile for _both_ SVE2p1 and SME2, the builtin can be used in either
  mode (non-streaming, streaming or streaming-compatible)

sdesmalen-arm committed:
[Clang][AArch64] Generalise streaming mode checks for builtins.

PR #76975 added 'IsStreamingOrSVE2p1' to emit a diagnostic when a builtin marked
with 'IsStreamingOrSVE2p1' is used in a non-streaming function that is not
compiled with `+sve2p1`.

The problem is a bit more complex than only this case. For example, we've marked
lots of builtins with 'IsStreamingCompatible', meaning it can be used in either
streaming, streaming-compatible or non-streaming functions. But the code in
SemaChecking, doesn't check the appropriate target guards. This issue becomes
relevant when SVE builtins are only available in streaming mode, e.g. when
compiling for SME without SVE.

If we were to add the appropriate target guards, we'd have to add many more
combinations, e.g.:
  IsStreamingSMEOrSVE
  IsStreamingSME2OrSVE2
  IsStreamingSMEOrSVE2p1
  IsStreamingSME2OrSVE2p1
  etc.

To avoid having to add more combinations (and avoid having to add more in the
future for new extensions), we use a single 'IsSVEOrStreamingSVE' flag for all
builtins that are available in streaming mode for the appropriate SME flags, or
in non-streaming mode for the appropriate SVE flags, or both.  The code in
SemaChecking will then verify for which mode (or both) the builtin would be
defined, given the target features of the function/compilation unit.

For example:

  'svclamp' is enabled under FEAT_SVE2p1 and FEAT_SME2

* When we compile for SVE2p1 and SME (but not SME2), the builtin is undefined
  behaviour when called from a streaming function.

* When we compile for SME2 and SVE2 (but not SVE2p1), the builtin is undefined
  behaviour when called from a non-streaming function.

* When we compile for _both_ SVE2p1 and SME2, the builtin can be used in either
  mode (non-streaming, streaming or streaming-compatible)

sdesmalen-arm committed:
Fix test failures

This changes the target guards for a small number of intrinsics, which is
the minimum set to fix the test Sema/CodeGen failures introduced by this
patch. I had hoped I could leave these changes to a follow-up patch to keep
this patch simple.

My next patch will set the correct target guards for *all* the intrinsics
(with corresponding RUN lines for the tests), which should complete the
work on the Clang side.

aemerson commented on clang/test/Sema/aarch64-streaming-sme-or-nonstreaming-sve-builtins.c:
```diff
@@ -0,0 +1,51 @@
+  return svclamp(a, b, c);
+}
+
+__attribute__((target("+sve2p1,+sme2")))
+svfloat32_t good3(svfloat32_t a, svfloat32_t b, svfloat32_t c) __arm_streaming_compatible {
+  return svclamp(a, b, c);
+}
```
For completeness can we have a test as well to check this case but instead with `__arm_streaming` as well

aemerson reviewed (approved):
LGTM with suggestion

sdesmalen-arm committed:
Add extra test-case

paulwalker-arm commented on clang/lib/Sema/SemaARM.cpp:
```diff
@@ -561,16 +561,61 @@ SemaARM::ArmStreamingType getArmStreamingFnType(const FunctionDecl *FD) {
+      CachedFD = FD;
+    }
+
+    if (SatisfiesSVE && SatisfiesSME)
```
Does this effectively prevent streaming compatible functions when only SVE feature flags are available?

My updated understand of streaming compatible functions is that SME features play no role and the user is expected to use SVE feature flags to direct the compiler to the level of SVE support a streaming compatible function can have, much like they would for ordinary functions.

paulwalker-arm commented on clang/include/clang/Basic/arm_sve_sme_incl.td:
```diff
@@ -225,7 +225,7 @@ def IsStreamingCompatible           : FlagType<0x4000000000>;
 def IsWriteZA                       : FlagType<0x10000000000>;
 def IsReductionQV                   : FlagType<0x20000000000>;
-def IsStreamingOrSVE2p1             : FlagType<0x40000000000>; // Use for intrinsics that are common between sme/sme2 and sve2p1.
+def IsSVEOrStreamingSVE             : FlagType<0x40000000000>; // Use for intrinsics that are common between SVE and SME.
```
If you permit a bit of bike shedding I don't think this is a good name.  From what I can see the new flag is used to trigger dynamic resolution to determine if the builtin is available to use based on the target features along with any keywords associated with the function.  Perhaps `RequiresDynamicVerification`? or ideally something shorter that has the same meaning.

sdesmalen-arm committed:
Address review comments

* Allow intrinsics in streaming-compatible functions, if they satisfy the
  SVE feature requirements.
* Renamed IsSVEOrStreamingSVE -> VerifyRuntimeMode

Additionally:

* Fixed caching mechanism; it previously cached the 'SatisfiesSVE' and
  'SatisfiesSME' values as well, which meant it cached per builtin call,
  not per functiondecl.
* As per offline discussion with @paulwalker-arm, I have changed warning
  for undefined behaviour into an error.

sdesmalen-arm committed:
Fix clang-format

sdesmalen-arm committed:
Address review comments.

This removes the invalid caching mechanism, that wouldn't be thread-safe.
Also returns directly when it finds a streaming-mode error, rather than using HasError.

sdesmalen-arm commented on clang/lib/Sema/SemaARM.cpp:
```diff
@@ -559,31 +559,86 @@ SemaARM::ArmStreamingType getArmStreamingFnType(const FunctionDecl *FD) {
+  // * When compiling for both SVE and SME, the caller can be in either mode.
+  if (BuiltinType == SemaARM::VerifyRuntimeMode) {
+    static llvm::StringMap<bool> CallerFeatureMapWithoutSVE,
+        CallerFeatureMapWithoutSME;
```
You're right, not sure what I was thinking here with the 'static' variable :)

I'd rather not go down the route of parsing the BuiltinTargetGuards here or making assumptions on the format if we're not going to ensure this format in the arm_sve.td file. For example, one can write `(sve,featureX)|(sme2,featureX)` or `(sve|sme),featureX`. Sure we can assume the former format, but in that case I think we need to have a `SVETargetGuard` and a `SMETargetGuard`, that we let TableGen combine into a canonical form for the combined `TargetGuard`.

sdesmalen-arm commented on clang/lib/Sema/SemaARM.cpp:
```diff
@@ -622,7 +679,8 @@ bool SemaARM::CheckSMEBuiltinFunctionCall(unsigned BuiltinID,
 
     if (BuiltinType)
-      checkArmStreamingBuiltin(SemaRef, TheCall, FD, *BuiltinType);
+      HasError |= checkArmStreamingBuiltin(SemaRef, TheCall, FD, *BuiltinType,
```
Done

sdesmalen-arm commented on clang/lib/Sema/SemaARM.cpp:
```diff
@@ -559,31 +559,86 @@ SemaARM::ArmStreamingType getArmStreamingFnType(const FunctionDecl *FD) {
+
+    // Cache the feature maps, to avoid having to recalculate this for each
+    // builtin call.
+    static unsigned CachedODRHash = 0;
```
Done

sdesmalen-arm committed:
Make FD const again

sdesmalen-arm committed:
Merge remote-tracking branch 'upstream/main' into generalise-streaming-mode-checks-builtins
